# Homework

The goal of this homework is to get familiar with MLflow, the tool for experiment tracking and model management

## Q1. Install MLflow

In [1]:
!mlflow --version

mlflow, version 2.17.0


**Answer**: The version of MLflow is 2.17.0

## Q2. Download and pre-process the data

In [15]:
import os
from pathlib import Path
import subprocess

data_root = Path("../data")
hw_02_data = data_root / "hw_02_data"

# Create hw_02_data when it doesn't exists
if not hw_02_data.is_dir():
    hw_02_data.mkdir()

# Run data preprocessing
if len(list(hw_02_data.glob("*.pkl"))) == 0:
    cmd = f"python homework/preprocess_data.py --raw_data_path {str(data_root)}/raw --dest_path {str(hw_02_data)}"
    subprocess.call(cmd, shell=True)

In [16]:
print(len(list(hw_02_data.glob("*.pkl"))))

4


**Answer**: There are 4 files

## Q3. Train a model with autolog

In [21]:
cmd = f"python homework/train.py --data_path {str(hw_02_data)}"
subprocess.call(cmd, shell=True)

2024/10/18 22:01:29 INFO mlflow.tracking.fluent: Experiment with name 'nyc-taxi-duration' does not exist. Creating a new experiment.
2024/10/18 22:01:29 INFO mlflow.tracking.fluent: Autologging successfully enabled for sklearn.
2024/10/18 22:01:30 WARNING mlflow.sklearn: Failed to log training dataset information to MLflow Tracking. Reason: 'numpy.ndarray' object has no attribute 'toarray'
/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


0

**Answer** The value of `min_samples_split` is 2

## Q.4 Launch the tracking server locally

**Answer** `default-artifact-root` is the option to configure `artifacts`

## Q.5 Tune model hyperparameter

In [26]:
cmd = f"python homework/hpo.py --data_path {str(hw_02_data)}"
subprocess.call(cmd, shell=True)

2024/10/18 22:18:40 INFO mlflow.tracking.fluent: Experiment with name 'random-forest-hyperopt' does not exist. Creating a new experiment.


  0%|          | 0/15 [00:00<?, ?trial/s, best loss=?]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:44 INFO mlflow.tracking._tracking_service.client: 🏃 View run abundant-stoat-606 at: http://127.0.0.1:5000/#/experiments/1/runs/a6bec016155f48a5abe21dace3303246.

2024/10/18 22:18:44 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



  7%|▋         | 1/15 [00:04<01:01,  4.41s/trial, best loss: 5.370086069268862]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:45 INFO mlflow.tracking._tracking_service.client: 🏃 View run luxuriant-mink-589 at: http://127.0.0.1:5000/#/experiments/1/runs/9e7ba02fc0da42b8940ddb469a83e90f.

2024/10/18 22:18:45 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 13%|█▎        | 2/15 [00:04<00:26,  2.01s/trial, best loss: 5.370086069268862]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:45 INFO mlflow.tracking._tracking_service.client: 🏃 View run defiant-ox-263 at: http://127.0.0.1:5000/#/experiments/1/runs/765fe9941b4445c29df54ecaa9470844.

2024/10/18 22:18:45 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 20%|██        | 3/15 [00:05<00:15,  1.29s/trial, best loss: 5.370086069268862]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:48 INFO mlflow.tracking._tracking_service.client: 🏃 View run serious-seal-66 at: http://127.0.0.1:5000/#/experiments/1/runs/06d2f23bfc884b26b5ebcedfe7c71f0e.

2024/10/18 22:18:48 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 27%|██▋       | 4/15 [00:08<00:22,  2.07s/trial, best loss: 5.357490752366866]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:50 INFO mlflow.tracking._tracking_service.client: 🏃 View run amusing-crow-932 at: http://127.0.0.1:5000/#/experiments/1/runs/5d185ef78c2342cf93824e87df09b38e.

2024/10/18 22:18:50 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 33%|███▎      | 5/15 [00:09<00:18,  1.84s/trial, best loss: 5.357490752366866]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:55 INFO mlflow.tracking._tracking_service.client: 🏃 View run capable-crane-859 at: http://127.0.0.1:5000/#/experiments/1/runs/d9475e12b76c450e80f57f38f94d250f.

2024/10/18 22:18:55 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 40%|████      | 6/15 [00:14<00:25,  2.81s/trial, best loss: 5.354700855292386]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:59 INFO mlflow.tracking._tracking_service.client: 🏃 View run bouncy-foal-675 at: http://127.0.0.1:5000/#/experiments/1/runs/dc956549af8444f8b6c2be82385b6f56.

2024/10/18 22:18:59 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 47%|████▋     | 7/15 [00:18<00:25,  3.19s/trial, best loss: 5.354700855292386]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:18:59 INFO mlflow.tracking._tracking_service.client: 🏃 View run rogue-moose-248 at: http://127.0.0.1:5000/#/experiments/1/runs/742a911f3f3f437bb7998d45ab9aa2dc.

2024/10/18 22:18:59 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 53%|█████▎    | 8/15 [00:19<00:16,  2.34s/trial, best loss: 5.354700855292386]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:02 INFO mlflow.tracking._tracking_service.client: 🏃 View run gregarious-conch-346 at: http://127.0.0.1:5000/#/experiments/1/runs/afd4c64b2d9c498a94fd37732310f19b.

2024/10/18 22:19:02 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 60%|██████    | 9/15 [00:21<00:15,  2.50s/trial, best loss: 5.354700855292386]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:04 INFO mlflow.tracking._tracking_service.client: 🏃 View run adaptable-horse-472 at: http://127.0.0.1:5000/#/experiments/1/runs/065dffa2e1d7485cab9a731d6e6a2c8c.

2024/10/18 22:19:04 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 67%|██████▋   | 10/15 [00:24<00:12,  2.45s/trial, best loss: 5.354700855292386]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:06 INFO mlflow.tracking._tracking_service.client: 🏃 View run debonair-fowl-225 at: http://127.0.0.1:5000/#/experiments/1/runs/9bb128a7e7394a908fba8106f460ff75.

2024/10/18 22:19:06 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 73%|███████▎  | 11/15 [00:25<00:08,  2.20s/trial, best loss: 5.335419588556921]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:08 INFO mlflow.tracking._tracking_service.client: 🏃 View run orderly-gnat-934 at: http://127.0.0.1:5000/#/experiments/1/runs/7723be5d458d45b781fed0918cc2032c.

2024/10/18 22:19:08 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 80%|████████  | 12/15 [00:27<00:06,  2.02s/trial, best loss: 5.335419588556921]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:08 INFO mlflow.tracking._tracking_service.client: 🏃 View run dazzling-carp-843 at: http://127.0.0.1:5000/#/experiments/1/runs/018b1b0b9a8b46e89a2dcc47cd2a0b38.

2024/10/18 22:19:08 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 87%|████████▋ | 13/15 [00:28<00:03,  1.59s/trial, best loss: 5.335419588556921]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:10 INFO mlflow.tracking._tracking_service.client: 🏃 View run popular-owl-648 at: http://127.0.0.1:5000/#/experiments/1/runs/8c39a144047b4bee96d64fcc4dc289b2.

2024/10/18 22:19:10 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



 93%|█████████▎| 14/15 [00:29<00:01,  1.57s/trial, best loss: 5.335419588556921]

/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(

2024/10/18 22:19:12 INFO mlflow.tracking._tracking_service.client: 🏃 View run receptive-hound-88 at: http://127.0.0.1:5000/#/experiments/1/runs/abbe544e27044316b2d76ffde506a4c9.

2024/10/18 22:19:12 INFO mlflow.tracking._tracking_service.client: 🧪 View experiment at: http://127.0.0.1:5000/#/experiments/1.



100%|██████████| 15/15 [00:31<00:00,  2.13s/trial, best loss: 5.335419588556921]


0

**Answer** The best RMSE is 5.335

## Q.6 Promote the best model to the model registry

In [34]:
cmd = f"python homework/register_model.py --data_path {str(hw_02_data)}"
subprocess.call(cmd, shell=True)

2024/10/18 23:08:54 WARNING mlflow.sklearn: Failed to log training dataset information to MLflow Tracking. Reason: 'numpy.ndarray' object has no attribute 'toarray'
/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
/Users/tnhgiang/code/personal/mlops-zoomcamp/env/lib/python3.11/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
2024/10/18 23:09:00 INFO mlflow.tracking._tracking_service.client: 🏃 View run sincere-shoat-360 at: http://127.0.0.1:5000/#/experiments/2/runs/0be08e25105046278382562533f53aad.
2024/10/18 23:09:00 INFO mlflow.tracking._tracking_service.cli

0

**Answer** The test RMSE of the best model is 5.567

In [28]:

from mlflow.entities import ViewType
from mlflow.tracking import MlflowClient
client = MlflowClient("http://127.0.0.1:5000/")

runs = client.search_runs(
    experiment_ids=1,
    run_view_type=ViewType.ACTIVE_ONLY,
    max_results=5,
    order_by=["metrics.rmse ASC"]
)

In [31]:
runs[0].info.run_id

'9bb128a7e7394a908fba8106f460ff75'